src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/CallArranger.java

```diff
@@ -270,10 +268,11 @@ protected BindingCalculator(boolean forArguments) {
                     while (offset < layout.byteSize()) {
                         final long copy = Math.min(layout.byteSize() - offset, 8);
                         VMStorage storage = regs[regIndex++];
-                        Class<?> type = SharedUtils.primitiveCarrierForSize(copy);
                         if (offset + copy < layout.byteSize()) {
                             bindings.dup();
                         }
+                        boolean useFloat = storage.type() == StorageClasses.VECTOR;
```

Comment:
So , my understanding here is that we had issues all along - e.g. this is unrelated to the upcall work?

Reply:
Not in issue so far since we were using an intermediate buffer. We simply write the bits as an integer, but the native code reads it as a long.

Since the specialized version doesn't use an intermediate buffer though, the types have to be aligned properly.